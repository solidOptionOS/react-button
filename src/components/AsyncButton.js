import React from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils';

export default class AsyncButton extends React.Component {
  state = {
    asyncState: null,
  };

  componentWillUnmount() {
    this.isUnmounted = true;
  }

  resetState() {
    this.setState({
      asyncState: null,
    });
  }

  handleClick(...args) {
    const clickHandler = this.props.onClick;
    if (typeof clickHandler === 'function') {
      this.setState({
        asyncState: 'pending',
      });

      const returnFn = clickHandler.apply(null, args);
      if (returnFn && typeof returnFn.then === 'function') {
        returnFn
          .then(() => {
            if (this.isUnmounted) {
              return;
            }
            this.setState({
              asyncState: 'fulfilled',
            });
          })
          .catch(error => {
            if (this.isUnmounted) {
              return;
            }
            this.setState({
              asyncState: 'rejected',
            });
            throw error;
          });
      } else {
        this.resetState();
      }
    }
  }

  render() {
    const {
      children,
      text,
      pendingText,
      fulFilledText,
      rejectedText,
      className = '',
      loadingClass,
      fulFilledClass,
      rejectedClass,
      disabled,
      ...attributes
    } = this.props;

    const { asyncState } = this.state;
    const isPending = asyncState === 'pending';
    const isFulfilled = asyncState === 'fulfilled';
    const isRejected = asyncState === 'rejected';
    const isDisabled = disabled || isPending;
    let buttonText;

    if (isPending) {
      buttonText = pendingText;
    } else if (isFulfilled) {
      buttonText = fulFilledText;
    } else if (isRejected) {
      buttonText = rejectedText;
    }
    buttonText = buttonText || text;

    return (
      <button
        {...attributes}
        className={classNames(className, {
          [loadingClass]: isPending,
          [fulFilledClass]: isFulfilled,
          [rejectedClass]: isRejected,
        })}
        disabled={isDisabled}
        onClick={event => this.handleClick(event)} 
      >
        {typeof children === 'function'
          ? children({
              buttonText,
              isPending,
              isFulfilled,
              isRejected,
            })
          : children || buttonText}
      </button>
    );
  }
}

AsyncButton.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  className: PropTypes.string,
  loadingClass: PropTypes.string,
  fulFilledClass: PropTypes.string,
  rejectedClass: PropTypes.string,
  disabled: PropTypes.bool,
  text: PropTypes.string,
  pendingText: PropTypes.string,
  fulFilledText: PropTypes.string,
  rejectedText: PropTypes.string,
  onClick: PropTypes.func,
};

AsyncButton.defaultProps = {
  loadingClass: 'AsyncButton--loading',
  fulFilledClass: 'AsyncButton--fulfilled',
  rejectedClass: 'AsyncButton--rejected',
};
