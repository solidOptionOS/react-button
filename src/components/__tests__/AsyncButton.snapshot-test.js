import React from 'react';
import AsyncButton from '../AsyncButton';
import renderer from 'react-test-renderer';

const expectToMatchSnapshot = component => {
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
}; 

describe('<AsyncButton /> Snapshots', () => {
  it('should have sane defaults', () => {
    expectToMatchSnapshot(<AsyncButton />);
  });

  it('should delegate text, className & disabled props', () => {
    expectToMatchSnapshot(
      <AsyncButton className="btn btn-default" text="Save" disabled={true} />
    );
  });
});

describe('Block form :- <AsyncButton></AsyncButton>', () => {
  it('should have sane defaults', () => {
    expectToMatchSnapshot(<AsyncButton>Save</AsyncButton>);
  });

  it('should delegate text, className & disabled props', () => {
    expectToMatchSnapshot(
      <AsyncButton className="btn btn-default" text="Save" disabled={true}>
        {({ buttonText, isPending }) =>
          <span>
            {buttonText}
          </span>}
      </AsyncButton>
    );
  });
});
