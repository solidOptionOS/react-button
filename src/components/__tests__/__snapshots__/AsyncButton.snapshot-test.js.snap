// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncButton /> Snapshots should delegate text, className & disabled props 1`] = `
<button
  className="btn btn-default"
  disabled={true}
  onClick={[Function]}
>
  Save
</button>
`;

exports[`<AsyncButton /> Snapshots should have sane defaults 1`] = `
<button
  className=""
  disabled={false}
  onClick={[Function]}
/>
`;

exports[`Block form :- <AsyncButton></AsyncButton> should delegate text, className & disabled props 1`] = `
<button
  className="btn btn-default"
  disabled={true}
  onClick={[Function]}
>
  <span>
    Save
  </span>
</button>
`;

exports[`Block form :- <AsyncButton></AsyncButton> should have sane defaults 1`] = `
<button
  className=""
  disabled={false}
  onClick={[Function]}
>
  Save
</button>
`;
